<h2>Run summary</h2>

<p>
  Subsequent sections of the report use the following to identify figures and
  other information.
</p>
@SAMPLE_KEY@

<p>
  Read counts. Filtered and aligned read counts are reported relative
  to the total number of reads (clusters; if only filtered or aligned
  reads are available, total read count is reported). Consult Genome
  Analyzer documentation for official guidelines. From experience,
  very good runs of the Genome Analyzer 'control' lane result in 25-30
  million reads, with up to 95&#37; passing pre-defined filters.
</p>
<pre>
  ShortRead:::.plotReadCount(qa)
</pre>
@PPN_COUNT@

<p>
  Base call frequency over all reads. Base frequencies should accurately
  reflect the frequencies of the regions sequenced.
</p>
<pre>
  ShortRead:::.plotNucleotideCount(qa)
</pre> 
@BASE_CALL_COUNT@

<p>
  Overall read quality. Lanes with consistently good quality reads have
  strong peaks at the right of the panel.
</p>
<pre>
  df &lt;- qa[["readQualityScore"]]
  ShortRead:::.plotReadQuality(df[df$type=="read",])
</pre>
@READ_QUALITY_FIGURE@
